docs/proposals/20221611-workload-identity-integration.md

```diff
@@ -0,0 +1,351 @@
+
+Also a go file e.g `workload_identity.go` in the `identity` package dealing with AWI functionality.
+
+```go
+
+type workloadIdentityCredential struct {
+	assertion string
+	file      string
+	cred      *azidentity.ClientAssertionCredential
+	lastRead  time.Time
+}
+
+type workloadIdentityCredentialOptions struct {
+	azcore.ClientOptions
+}
+
+func newWorkloadIdentityCredential(tenantID, clientID, file string, options *workloadIdentityCredentialOptions) (*workloadIdentityCredential, error) {
+	w := &workloadIdentityCredential{file: file}
+	cred, err := azidentity.NewClientAssertionCredential(tenantID, clientID, w.getAssertion, &azidentity.ClientAssertionCredentialOptions{ClientOptions: options.ClientOptions})
+	if err != nil {
+		return nil, err
+	}
+	w.cred = cred
+	return w, nil
+}
+
+func (w *workloadIdentityCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return w.cred.GetToken(ctx, opts)
+}
+
+func (w *workloadIdentityCredential) getAssertion(context.Context) (string, error) {
+	if now := time.Now(); w.lastRead.Add(5 * time.Minute).Before(now) {
+		content, err := os.ReadFile(w.file)
+		if err != nil {
+			return "", err
+		}
+		w.assertion = string(content)
+		w.lastRead = now
+	}
+	return w.assertion, nil
+}
+
+```
+
+
+### <a name='OpenQuestions'></a>Open Questions
+
+#### <a name='Howtomultitenancy'></a>1. How to achieve multi-tenancy?
+
+The identity is tied to service account via the client ID annotation present in the service account. Service account has one to one mapping with the pod. To have a many-to-one relationship between multiple identities and a service account, multiple federated identity credentials can be created that references the same service account.
+
+`azure.workload.identity/client-id` annotations contains the client ID of the identity that will be used.
+
+To use a different client ID, the client ID should be updated but the limitation is that at a time only one identity can be used. 
+
+
+#### <a name='Howtodistributekeys'></a>2. How to distribute key pair to management cluster?
+
+This an open question to discuss to find a better way to distribute keys to the control plane node for the management cluster.
+
+
+#### <a name='UserExperience'></a>3. User Experience
+
+Though AWI has a lot of advantages as compared to AWI, setting up AWI involves couple of manual step and it can impact the user experience. 
+
+### <a name='MigrationPlan'></a>Migration Plan
+
+Management clusters using AAD pod identity should have a seamless migration process which is well documented.
+
+As part of upgrade that supports AWI, the env variable can be set to true to use AWI. 
+But if capz uses AWI, it can use only one identity at a time and it can pose challenges to migration if the management cluster is using different identity simultaneously from the AAD pod identity. 
+
+### <a name='TestPlan'></a>Test Plan
```

Comment:
I'm very interested to learn more about how we can set this up for our CI sub. @aramase do you have examples of CI using kind clusters to test AZWI?

Reply:
Though Cecile is already aware of this. Pasting this issue link that outlines this for more curious readers 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2936